### PR TITLE
WIP: TYPE: Region folding

### LINKED ISSUE #2316: Region folding not supported
regionfoldinggif

#### Region Folding
Region folding is a wonderful code-organization feature standard in Intellij/Webstorm/Intellij-Scala etc. where blocks of code enclosed by region tags may be folded.
```
 //region * some name * 
* some code *
//endregion
```
Creating regions can be done manually or by making use of the standard 'Surround With' -> Region functionality.

#### Problem
Intellij/Jetbrains region folding is not currently supported. There's an option for it in Settings under Code Folding but it has no effect: 
img

Currently we can fold functions, mods, extern crates, use etc. which is a good start. The inability to do properly nested chunks of foldable code (with custom labels!) limits code navigation speed and organization for those proficient at it. 

A benefit of proper region folding over function/body folding is that it can reduce cognitive effort required when navigating code with long/tall(many lines even when 'folded') type signatures.

I spent a night trying to implement it (going so far as adding a parsing token for it and getting '//region' recognized as something similar to a block comment)...but my efforts fell short due to lack of Kotlin/codegeneration knowledge. Memory's a little fuzzy but I remember there being something in another custom language plugin that just used a standard com.intellij.something to add the support without much todo (Maybe don't need to monkey around with the parser to get it working?).

### PR DESCRIPTION
**Pull request by anton-okolelov**
fixes #2316

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/folding/RsFoldingBuilder.kt [hunk 1]
```diff
@@ -46,11 +47,12 @@ class RsFoldingBuilder : FoldingBuilderEx(), DumbAware {
             node.psi is PsiComment -> "/* ... */"
             else -> "{...}"
         }
+    }
 
-    override fun buildFoldRegions(root: PsiElement, document: Document, quick: Boolean): Array<out FoldingDescriptor> {
-        if (root !is RsFile) return emptyArray()
+    override fun buildLanguageFoldRegions(descriptors: MutableList<FoldingDescriptor>, root: PsiElement, document: Document, quick: Boolean) {
+        if (root !is RsFile) return;
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/folding/RsFoldingBuilder.kt]: `;` is redundant here

File: src/main/kotlin/org/rust/ide/folding/RsFoldingBuilder.kt [hunk 2]
```diff
@@ -46,11 +47,12 @@ class RsFoldingBuilder : FoldingBuilderEx(), DumbAware {
...
 
-        val descriptors: MutableList<FoldingDescriptor> = ArrayList()
+        //val descriptors: MutableList<FoldingDescriptor> = ArrayList()
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/folding/RsFoldingBuilder.kt]: Let's drop this line

File: src/main/kotlin/org/rust/ide/folding/RsFoldingBuilder.kt [hunk 3]
```diff
@@ -7,6 +7,7 @@ package org.rust.ide.folding
 
 import com.intellij.codeInsight.folding.CodeFoldingSettings
 import com.intellij.lang.ASTNode
+import com.intellij.lang.folding.CustomFoldingBuilder
 import com.intellij.lang.folding.FoldingBuilderEx
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/folding/RsFoldingBuilder.kt]: it's unused import now

File: src/test/resources/org/rust/ide/folding/fixtures/custom_region.rs [hunk 4]
```diff
@@ -0,0 +1,4 @@
+<fold text='custom text'>//region custom text
+//some code
```
Comment (Undin) [inline src/test/resources/org/rust/ide/folding/fixtures/custom_region.rs]: Maybe add some real code here instead of comment?

Comment (anton-okolelov) [reply src/test/resources/org/rust/ide/folding/fixtures/custom_region.rs]: Yeah, I'm working on it right now

Comment (anton-okolelov) [reply src/test/resources/org/rust/ide/folding/fixtures/custom_region.rs]: I run into some problems here

**Comment by anton-okolelov** [comment]
I didn't feel well these days, but now I'm okay and ready to continue working on it this evening.
By the way, I fixed minor issues you were mentioned earlier.

For the moment the only problem is when I run this test it doesn't work:

```
<fold text='custom text'>//region custom text
fn main() <fold text='{...}'>{
}</fold>
//endregion</fold>
```

but this works okay:
```
<fold text='custom text'>//region custom text

fn main() <fold text='{...}'>{
}</fold>
//endregion</fold>
```

the difference is empty line before `fn main`. Have no idea why, but hope today I'll figure out

**Comment by anton-okolelov** [comment]
It seems I have found the reason of this behavior. It seems there is a bug in parser.

For example:
rust file:
```rust
//region

fn main() {}
//endregion
```

PSI structure:
```
FILE(0,34)
  PsiComment(<EOL_COMMENT>)('//region')(0,8)
  PsiWhiteSpace('\n\n')(8,10)
  RsFunctionImpl(FUNCTION)(10,22)
    PsiElement(fn)('fn')(10,12)
    PsiWhiteSpace(' ')(12,13)
    PsiElement(identifier)('main')(13,17)
    RsValueParameterListImpl(VALUE_PARAMETER_LIST)(17,19)
      PsiElement(()('(')(17,18)
      PsiElement())(')')(18,19)
    PsiWhiteSpace(' ')(19,20)
    RsBlockImpl(BLOCK)(20,22)
      PsiElement({)('{')(20,21)
      PsiElement(})('}')(21,22)
  PsiWhiteSpace('\n')(22,23)
  PsiComment(<EOL_COMMENT>)('//endregion')(23,34)
```

And now without new line before function:
```rust
//region
fn main() {}
//endregion
```

PSI: 
```
FILE(0,33)
  RsFunctionImpl(FUNCTION)(0,21)
    PsiComment(<EOL_COMMENT>)('//region')(0,8)
    PsiWhiteSpace('\n')(8,9)
    PsiElement(fn)('fn')(9,11)
    PsiWhiteSpace(' ')(11,12)
    PsiElement(identifier)('main')(12,16)
    RsValueParameterListImpl(VALUE_PARAMETER_LIST)(16,18)
      PsiElement(()('(')(16,17)
      PsiElement())(')')(17,18)
    PsiWhiteSpace(' ')(18,19)
    RsBlockImpl(BLOCK)(19,21)
      PsiElement({)('{')(19,20)
      PsiElement(})('}')(20,21)
  PsiWhiteSpace('\n')(21,22)
  PsiComment(<EOL_COMMENT>)('//endregion')(22,33)
```

Comment is inside RsFunctionImpl now. As for me it's a little bit weird.

@Undin or  @matklad , could you clarify, Is it a bug or a feature?  Should I fix parser?

**Comment by anton-okolelov** [comment]
I have modified ADJACENT_LINE_COMMENTS field to fix this issue. Not sure if it is the correct way to do that

**Comment by anton-okolelov** [comment]
Could anybody do code review?

**Comment by anton-okolelov** [comment]
Thank you for the response
It make sense, I'm going to try to figure out what happens in java

**Comment by adrianperreault** [comment]
Any luck with it @anton-okolelov or are you throwing in the towel on this one?

**Comment by anton-okolelov** [comment]
I tried to figure out what's happen in java, but did not succeed. To be honest, I just have not enough time for this. 
But I started this PR and I feel that I must to finish this.
I promise to try again next weekend.

**Comment by anton-okolelov** [comment]
this feature has already implemented in #3005

**Pull request closed without merging**